Used Lab 3 GWAS code as a reference to help us implement our linear regression model!

# Citations
1. Massarat, A. R., Lamkin, M., Reeve, C., Williams, A. L., D’Antonio, M., & Gymrek,
        M. Haptools: a toolkit for admixture and haplotype analysis [Computer software].
        https://github.com/CAST-genomics/haptools who created haptools 
        (Was granted permission to use from Professor Gymrek)
2. Purcell S, Neale B, Todd-Brown K, Thomas L, Ferreira MAR, Bender D, Maller J, de Bakker PIW, Daly MJ & Sham PC (in press) PLINK: a       toolset for whole-genome association and population-based linkage analysis. American Journal of Human Genetics.
Gymrek, M. Lab 3 Population genetics and GWAS.
3. The 1000 Genomes Project Consortium. A global reference for human genetic variation. Nature 526, 68–74 (2015). https://doi.org/10.1038/nature15393
4. Ramos , Leodanis  Pozo. “Build Command-Line Interfaces With Python’s Argparse.” Real Python, 7 Dec. 2022, https://realpython.com/command-line-interfaces-python-argparse/.

## Python Packages Used
- Haptools
- Numpy
- Statsmodels
- Qqman  
- Argparse
- Pandas
- Matplotlib

# Resources Used
What plink does:
[https://zzz.bwh.harvard.edu/plink/]()

Things to work on:
- Switch to a linear regression---
- preprocesssing --
  - removing p values--
- Confounding factors?
- Generalizing code
- improving efficinecy


Presentation Notes
- Purpose/imputs/outputs
  - Accessibility? -- can pip install
- Add a references slide
- Challenges
