![python](https://img.shields.io/badge/python-3.9.5-green)
![matplotlib](https://img.shields.io/badge/matplotlib-3.4.2-red)
![numpy](https://img.shields.io/badge/numpy-1.21.1-blue)
![pandas](https://img.shields.io/badge/pandas-1.5.3-white)
![project_status](https://img.shields.io/badge/project__status-work%20in%20progress-orange)

# Blink-python-GWAS
This tool, Blink, is used to perform genome-wide association studies (GWAS). GWAS is important as it helps scientists identify genes associated with diseases or traits. Our implementation will compared to a GWAS tool called plink which is well-known in the bioinformatic field. 

NOTE: WE ARE RUNNING OUR CODE MAINLY ON JUPYTER NOTEBOOK FOR THE TIME BEING
- You will unfortunately need to install a 2gb file that contains the genotypes from 1000Genomes to run the code on Jupyter Notebook
- Please upload that vcf.gz file to your directory
- Follow the instructions on the notebook

## Installation instructions
Installation requires the `haptools` library to be installed. You can install these with pip:

`pip install haptools`

Once the required libraries are installed you can install `blink` with the following command:

`python setup.py install` (to be implemented this coming week)

Note: If you do not have root access, you can run the commands aboce with the additional options to install locally:

`pip install --user haptools` or (`pip install haptools` --> `ls ~/.local/bin/` --> `export PATH=$PATH:$HOME/.local/bin`)

`python setup.py install --user`

If `blink` was installed correctly, try `blink --help` to see if instructions to run `blink` are provided.

## Basic usage instructions
Using the python/jupyter notebook files following the instructions in the cell blocks/comments to run the code. You will most likely need to fill in your corresponding files/data to run the code. FOLLOW THE INSTRUCTIONS PROVIDED!
* For the simulate_pheno jupyter notebook, you will need to use your own variants to run the tool for converting .vcf files to .hap files

The basic usage of `blink` will be: 

`blink [other options]` (will be specified soon once we integrate everything)

## Complete usage instructions 
We have no fully implemented the complete usage of `blink` yet and will be incorporating it this upcoming week. (The options below will not work properly yet)
* `-g FILE`, `--genotypes FILE` --> input a genotype file (.vcf.gz format or else an error will be thrown)
* `-h FILE`, `--haplotypes FILE` --> input a haplotype file (.hap format or else an error will be thrown)
* `-o FILE`, `--output FILE` --> specify an output file (for our current code a .pheno output file should be specify)

## Credits
This repository was generated by Abhishek Ganga, Daniyal Rahman, and Jung Tzen Liew with inspiration from plink, haptools, and the CSE 185 Lab 3 projects. Please submit a pull request with any corrections or suggestions! We accept all criticism.
